src/PlonkVerify.hs

```diff
@@ -1,27 +1,157 @@
 
-{-# OPTIONS_GHC -Wno-unused-imports #-}
+module PlonkVerify
+( verify_plonk_bn254
+, VKey(..)
+, Proof(..)
+, ProgramId(..)
+, PublicParameter(..)
+, PublicParameterHash(..)
+, mkPublicParameterHash
+, verifyPlonkBn254
+, verifyPlonkBn254'
+) where
 
-module PlonkVerify (verify_plonk_bn254) where
+import Data.ByteString qualified as B
+import Data.ByteString.Base16 qualified as B16
+import Data.ByteString.Short qualified as BS
 
-import Data.Int
-import Data.Word
 import Foreign.C.String
 import Foreign.C.Types
-import Foreign.Ptr
 
-foreign import ccall safe "__c_verify_plonk_bn254" verify_plonk_bn254 :: CString -> CString -> CString -> CString -> IO (CUInt)
\ No newline at end of file
+import Language.Haskell.TH.Syntax (Lift)
+
+import System.IO.Temp
+import Data.Bits
+
+-- -------------------------------------------------------------------------- --
+-- Plonk Verifier FFI
+
+foreign import ccall safe "__c_verify_plonk_bn254"
+    verify_plonk_bn254 :: CString -> CString -> CString -> CString -> IO (CUInt)
+
+foreign import ccall safe "__c_verify_plonk_bn254_hashed"
+    verify_plonk_bn254_hashed :: CString -> CString -> CString -> CString -> IO (CUInt)
+
+-- -------------------------------------------------------------------------- --
+-- API Types
+
+-- | The verifying key identifies the circuit of the VM that is used to execute
+-- and proof the program. Intuitively, one can think of the key as a version
+-- identifier of the VM.
+--
+-- When the VM circuit is updated the key changes. Applications should record
+-- what key a proofs depends on for verification. The module "EmbedVKey"
+-- provides tools for embedding existing keys from files into applications
+-- binaries.
+--
+newtype VKey = VKey
+    { _vKey :: B.ByteString }
+    deriving (Show, Eq, Ord)
+    deriving (Lift)
+
+-- | The actual proof. This is opaque cryptographic object that allows the
+-- verifier to establish that a program was executed with a particular list of
+-- public parameters in the context of the VM that is idendified with a
+-- particular verifying key.
+--
+newtype Proof = Proof
+    { _proof :: BS.ShortByteString }
+    deriving (Show, Eq, Ord)
+
+-- | A unique identifier of a program.
+--
+newtype ProgramId = ProgramId
+    { _programId :: BS.ShortByteString }
+    deriving (Show, Eq, Ord)
+
+-- | The public parameters of a program invocation.
+--
+newtype PublicParameter = PublicParameter
+    { _publicParameter :: BS.ShortByteString }
+    deriving (Show, Eq, Ord)
+
+-- | The hash of the public parameters of a program invocation.
+--
+newtype PublicParameterHash = PublicParameterHash
+    { _publicParameterHash :: BS.ShortByteString }
+    deriving (Show, Eq, Ord)
+
+mkPublicParameterHash :: BS.ShortByteString -> Either String PublicParameterHash
+mkPublicParameterHash bytes
+    | BS.length bytes /= 32 = Left $ "wrong length; expected: 32, actual: " <> show (BS.length bytes)
+    | BS.head bytes .&. 0xe0 /= 0 = Left "first three bit are not set to 0"
```

Comment:
It should be documented, at least with a comment, that this is field-specific. 

That is, this is a Plonk verification for Plonk instantiated on [BN254](https://hackmd.io/@jpw/bn254). Should this one day switch to a distinct pairing-equipped curve (e.g. BLS12-381), this would need changes.

Reply:
I addressed this comment by renaming the whole module into `PlonkBn254.Verify`, which makes it clear that currently everything is parameterized on that curve.